### PR TITLE
Add s3-sqs-lambda-terraform pattern

### PR DESCRIPTION
**Pull request by enmata**
*Issue #, if available:*
Adding s3-sqs-lambda-terraform pattern

*Description of changes:*
Adding the terraform version of Add s3-sqs-lambda pattern
https://serverlessland.com/patterns/s3-sqs-lambda?ref=search

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

[example-pattern.json](https://github.com/user-attachments/files/16124255/example-pattern.json)

### CODE REVIEW

**Comment by bfreiberg** [comment]
Hi @enmata and thanks for the PR. We already have two Terraform patterns that cover this behaviour: S3 -> SQS (https://serverlessland.com/patterns/s3-sqs-cdk-terraform) and SQS -> Lambda (https://serverlessland.com/patterns/sqs-lambda-terraform-python). 
As patterns are supposed to be small, independent buildings builds I don't see any additional value in this submission.

File: s3-sqs-lambda-terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+  
+#################################################################
+# Outputs
+#################################################################
+# Displaying the SQS Queue, SourceS3 buckets and Lambda Function
+output "SQSQueueName" {
+  value       = aws_sqs_queue.MyHandlerQueue.name
+  description = "SQS Queue for queuing the s3 events"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/main.tf:164]:
```suggestion
  description = "SQS queue for queuing the s3 events"
```

File: s3-sqs-lambda-terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,173 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 3.27"
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/main.tf:5]: Not in particular. Upgrading it.

File: s3-sqs-lambda-terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+#################################################################
+# Lambda Function
+#################################################################
+# Creating Lambda Function
+resource "aws_lambda_function" "MyHandlerFunction-Function" {
+  filename      = data.archive_file.LambdaZipFile.output_path
+  function_name = "s3-sqs-lambda-tf-LambdaFunction"
+  role          = aws_iam_role.MyHandlerFunction-Role.arn
+  handler       = "app.handler"
+  runtime       = "nodejs16.x"
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/main.tf:87]: Upgrading it

File: s3-sqs-lambda-terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  value       = aws_sqs_queue.MyHandlerQueue.name
+  description = "SQS Queue for queuing the s3 events"
+}
+output "SourceS3BucketName" {
+  value       = aws_s3_bucket.MySourceS3Bucket.id
+  description = "S3 Bucket for object storage"
+}
+output "LambdaFunctionArn" {
+  value       = aws_lambda_function.MyHandlerFunction-Function.arn
+  description = "ResizerFunction function Arn"
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/main.tf:172]: Due CDK version of the pattern. Renaming it to HandlerFunction

File: s3-sqs-lambda-terraform/src/app.mjs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+console.log('Loading function');
+
+export const handler = async (event, context) => {
+    //console.log('Received event:', JSON.stringify(event, null, 2));
+    console.log('value1 =', event.key1);
+    console.log('value2 =', event.key2);
+    console.log('value3 =', event.key3);
+    return event.key1;  // Echo back the first key value
+    // throw new Error('Something went wrong');
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/src/app.mjs:9]: I Agree, removing it

File: s3-sqs-lambda-terraform/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
+# AWS Amazon S3 Lambda with SQS queue
```
[the remark this answers is not in the document]
Comment (bfreiberg) [reply s3-sqs-lambda-terraform/README.md:1]: There is a "to" missing after S3

File: s3-sqs-lambda-terraform/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+The terraform manifest deploys a Lambda function, an SQS queue, one S3 buckets and the IAM resources required to run the application. Create a Lambda function that is triggered on every new jpg image file stored on s3 Bucket using SQS as a notification target. The Lambda app acts as a Boilerplate code.
```
[the remark this answers is not in the document]
Comment (bfreiberg) [reply s3-sqs-lambda-terraform/README.md:3]: I don't see this changed in the latest commit, did you forget to include it?

File: s3-sqs-lambda-terraform/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    ```
+1. During the prompts:
+    * Enter yes
+1. Note the outputs from the deployment process. These contain the resource names and/or ARNs which are used for testing.
+
+
+## How it works
+
+* Use the AWS CLI or AWS console to upload an image to the source S3 Bucket
+* If the object is a .jpg the code, a Lambda that is triggered using SQS as a notification target.
```
[the remark this answers is not in the document]
Comment (bfreiberg) [reply s3-sqs-lambda-terraform/README.md:44]: I don't see this changed in the latest commit, did you forget to include it?

**Comment by enmata** [comment]
Sorry, I forget to add comments changes. Now its updated.

File: s3-sqs-lambda-terraform/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/README.md]: Sorry, updated.

File: s3-sqs-lambda-terraform/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+
+## How it works
+
+* Use the AWS CLI or AWS console to upload an image to the source S3 Bucket
+* If the object is a .jpg the code, a Lambda that is triggered using SQS as a notification target. 
+
+## Testing
+
+Run the following S3 CLI command to upload an image to the S3 bucket. Note, you must edit the {SourceBucketName} placeholder with the name of the source S3 Bucket. This is provided in the stack outputs.
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/README.md]: Sorry, updated.

File: s3-sqs-lambda-terraform/README.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+```bash
+aws s3 cp './events/exampleImage.png'  s3://{SourceBucketName}
+```
+
+## Documentations and next step
+
+To create a full Step Functions workflow with the pattern created, you can find out example workflows at Step Functions Workflow: [serverlessland.com/workflows](https://serverlessland.com/workflows)
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/README.md]: Sorry, updated.

File: s3-sqs-lambda-terraform/README.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
+# Amazon S3 to Amaozn SQS queue to AWS Lambda
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/README.md:1]:
```suggestion
# Amazon S3 to Amazon SQS queue to AWS Lambda
```

**Review by bfreiberg — COMMENTED** [review]
Looks good, just one typo left. However, you are missing the `example-pattern.json`.

> 
> 1. Use the model template located at https://github.com/aws-samples/serverless-patterns/tree/main/_pattern-model to set up a README, template and any associated code.
> 
> 2. THIS PROCESS HAS BEEN SIMPLIFIED. All the information below must be provided in the "example-pattern.json" file cloned from the model **

The above text is taken from the new pattern template.

File: s3-sqs-lambda-terraform/README.md (hunk 11 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/README.md]: Typo fixed, thanks.

example-pattern.json file added as an attachment on initial post on these PR. Do you want me do add it on the commit?

**Comment by bfreiberg** [comment]
Thanks for the update. To be clear, the example-pattern.json file must be added to the PR and filled out by you with corresponding values, e.g. language, level, relevant links etc. 
See [here](https://github.com/aws-samples/serverless-patterns/pull/2267/files#diff-6c2d94517527769d053ad15944429ac923262ae05cd2a6538cc09d65973553d5) for an example from a different pattern.

**Comment by enmata** [comment]
Added example-pattern.json

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 12]
```diff
@@ -0,0 +1,61 @@
+{
+  "title": "S3 to Step Functions using EventBridge",
+  "description": "How to create an EventBridge rule with S3 as the event source and Step Functions as target.",
+  "language": "nodejs",
+  "level": "200",
+  "framework": "Terraform",
+  "introBox": {
+    "headline": "How to create an EventBridge rule with S3 as the event source and Step Functions as target.",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:8]: Please update to match the actual pattern

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    "text": [
+      "This sample project demonstrates how you can trigger an execution in Step Functions state machine when an object is created in an S3 bucket.",
+      "When you uploads an object to the newly created S3 bucket, this will send an `Object Created` event to EventBridge, based on the EventBridge rule, the state machine is executed.",
+      "This pattern deploys one Step Functions state machine, one S3 bucket, one EventBridge rule."
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:12]: see above

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "AWS S3 Event Notifications with EventBridge",
+        "link": "https://aws.amazon.com/blogs/aws/new-use-amazon-s3-event-notifications-with-amazon-eventbridge"
+      },
+      {
+        "text": "Using EventBridge with S3",
+        "link": "https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventBridge.html"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:31]: Please update with relevant links

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    ]
+  },
+  "testing": {
+    "text": [
+      "See the GitHub repo for detailed testing instructions."
+    ]
+  },
+  "cleanup": {
+    "text": [
+      "terraform destroy"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:48]:
```suggestion
      "<code>terraform destroy</code>"
```

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    ]
+  },
+  "authors": [
+    {
+      "name": "Oriol Matavacas",
+      "image": "https://togithub.s3.eu-west-1.amazonaws.com/Oriol.jpg",
+      "bio": "Oriol Matavacas is a Sr. Solutions Architect at AWS based in Barcelona. Oriol primarily supporting customers on the journey to the Cloud. He enjoys building new solutions with scalability, availability and easy to maintain by using serverless.",
+      "linkedin": "https://www.linkedin.com/in/oriol-matavacas-rodriguez-b165868a",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:56]:
```suggestion
      "linkedin": "oriol-matavacas-rodriguez-b165868a",
```

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 17]
```diff
@@ -0,0 +1,61 @@
+{
+  "title": "S3 to Step Functions using EventBridge",
+  "description": "How to create an EventBridge rule with S3 as the event source and Step Functions as target.",
+  "language": "nodejs",
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: ok, changed

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 18]
```diff
@@ -0,0 +1,61 @@
+{
+  "title": "S3 to Step Functions using EventBridge",
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Sorry bout the confusion, updated.

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+  "description": "How to create an EventBridge rule with S3 as the event source and Step Functions as target.",
```
[the remark this answers is not in the document]
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Sorry bout the confusion, updated.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 12 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json:8]: Sorry bout the confusion, updated.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 13 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Sorry bout the confusion, updated.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 14 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Done

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 15 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Good point, all code commands updated

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 16 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Done

**Comment by enmata** [comment]
Thanks for the comments and sorry bout the confusion. Now its updated.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 12 above)
Comment (bfreiberg) [reply s3-sqs-lambda-terraform/example-pattern.json:8]: This still seems to be unresolved

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+  "description": "Create a Lambda function triggered for every jpg image file stored to S3 via SQS.",
+  "language": "node.js",
+  "level": "200",
+  "framework": "Terraform",
+  "introBox": {
+    "headline": "How to create an EventBridge rule with S3 as the event source and Step Functions as target.",
+    "text": [
+      "The terraform manifest deploys a Lambda function, an SQS queue, one S3 bucket and the IAM resources required to run the application.",
+      "An SQS queue consumes ObjectCreated events from an Amazon S3 bucket if the file has .jpg extension. The SQS triggers a Lambda function.",
+      "The Lambda app acts as a boilerplate code.",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:12]: What do you mean by this?

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Building resilient serverless patterns by combining messaging services",
+        "link": "https://github.com/aws-samples/s3-to-lambda-patterns"
+      },
+      {
+        "text": "S3-to-Lambda serverless applications",
+        "link": "https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventBridge.html"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:31]: Link and text don't seem to match

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+      "templateURL": "serverless-patterns/s3-sqs-lambda-terraform",
+      "projectFolder": "s3-sqs-lambda-terraform",
+      "templateFile": "main.tf"
+    }
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Building resilient serverless patterns by combining messaging services",
+        "link": "https://github.com/aws-samples/s3-to-lambda-patterns"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:27]: Link is outdated and doesn't really match the text

Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Thanks for the heads up, reference updated

**Comment by enmata** [comment]
Thanks for the comments, now updated its updated again.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 21 above)
Comment (bfreiberg) [reply s3-sqs-lambda-terraform/example-pattern.json:31]: Please note that this is still resolved

Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Updated

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 20 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Explanation updated

**Comment by enmata** [comment]
Changes applied. Please check when you have a moment.

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 23]
```diff
@@ -0,0 +1,61 @@
+{
+  "title": "AWS Amazon S3 Lambda with SQS queue",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:2]:
```suggestion
  "title": "Amazon S3 to AWS Lambda with Amazon SQS queue",
```

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+  "description": "Create a Lambda function triggered for every jpg image file stored to S3 via SQS.",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:3]:
```suggestion
  "description": "Create a Lambda function that is triggered for every jpg image file uploaded to S3 via an SQS queue.",
```

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+  "language": "node.js",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:4]:
```suggestion
  "language": "Node.js",
```

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+  "level": "200",
+  "framework": "Terraform",
+  "introBox": {
+    "headline": "Create a Lambda function triggered for every jpg image file stored to S3 via SQS",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:8]: The headline needs to stay below 75 characters, please reword

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+    }
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Configuring an Amazon SQS queue to trigger an AWS Lambda function",
+        "link": "https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-configure-lambda-function-trigger.html",
+      },
+      {
+        "text": "Amazon Simple Storage Service (S3) -> User Guide -> Using EventBridge",
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:30]: How does EventBridge relate to the rest of the pattern? It isn't mentioned anywhere else

File: s3-sqs-lambda-terraform/example-pattern.json [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+        "link": "https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventBridge.html"
+      }
+    ]
+  },
+  "deploy": {
+    "text": [
+      "<code>terraform init</code>",
+      "<code>terraform apply</code>"
```
Comment (bfreiberg) [inline s3-sqs-lambda-terraform/example-pattern.json:38]: <code> tags can be removed here as they are automatically added to the `deploy` field

**Review by bfreiberg — CHANGES_REQUESTED** [review]

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 24 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Thanks, changes applied

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 25 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Uppercase now

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 26 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json:8]: Compressed and resumed now. Please check

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 27 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Right and sorry for that. Reference deleted.

File: s3-sqs-lambda-terraform/example-pattern.json (hunk 28 above)
Comment (enmata) [reply s3-sqs-lambda-terraform/example-pattern.json]: Corrected

**Review by bfreiberg — APPROVED** [review]

**Pull request merged**